Adding some server rendering unit tests.

aickin opened the pull request:
This PR adds a handful of server rendering unit tests.

With the move to Fiber, it's likely that the server rendering path and client rendering path are going to diverge significantly, and I think it'd be useful to have a more robust unit test suite for server rendering as the work on Fiber SSR begins.

This PR starts to port over tests that I wrote in (since abandoned) PR #6836. This PR only ports over some of the helper functions and 4 of the simplest tests. If this PR is accepted, I plan to submit the rest of the tests in a series of further PRs; I only added four tests here because I wanted to keep this PR as digestible as possible.

Thanks for all your great work!

Note: I ran tests, linting, and flow, but I could not get `./scripts/fiber/record-tests` to work. It first complained about not being able to find `jest-cli`; when I npm installed jest-cli, it complained about an unresolved Promise with the error `TypeError: Path must be a string. Received undefined`.  I expect the CI build will fail on this, and I'd appreciate any pointers to getting it to work.

aickin commented:
Weirdly, the CI build seems to have succeeded, even though I haven't updated the list of failing fiber tests. I would greatly appreciate any help from the core team in figuring out what's going on there, and in figuring out how to run the update script on my machine.

Simultaneously, I've ported the rest of my SSR tests in a separate local branch; there are about 550 extra tests in total, and they are all passing on Stack. Unsurprisingly, many fail on Fiber, but that's arguably a good thing! If and when this PR is merged, I'll file a series of PRs to add more tests.

spicyj commented:
@acdlite fixed the test runner a couple days ago in #9060 – a Jest upgrade broke it, nothing you did. We're not sure why CI didn't catch that. If you rebase over that then it should work locally.

I wouldn't actually expect these tests to generally fail on Fiber. Do you have an example of a failing one handy?

I expect to have time to look at this tomorrow.

aickin commented:
@spicyj: Thanks for the tip on the test runner; I merged from master and was able to run the test list update scripts. Committed in 9257660.

> I wouldn't actually expect these tests to generally fail on Fiber.

All the `itRenders` tests are run four times in different rendering situations:

- Server render to string
- Client render on top of an empty DOM
- Client render on top of good server markup
- Client render on top of a bad server markup

The last of those (client render on bad markup) is expected to raise a markup mismatch warning, which Fiber can't do yet, so all those tests currently fail.

aickin commented:
> I expect to have time to look at this tomorrow.

Oh, and thanks for that! 👍

zackargyle commented:
I just want to throw out how excited I am about this. Thank you @aickin for doing this! Really stoked that Fiber SR is getting some attention. ✌🏻

spicyj commented on src/renderers/dom/shared/__tests__/ReactServerRendering-test.js:
```diff
@@ -23,20 +23,132 @@ var ReactTestUtils;
+//
+// Since all of the renders in this function are on the client, you can test interactivity,
+// unlike with itRenders.
+function itClientRenders(desc, testFn) {
```
Any reason these three are split from the other one?

aickin replied:
None of the tests in this PR use it, but some of the ones coming later do. 

The idea is that `itRenders` is for tests that only need to inspect the DOM, whereas `itClientRenders` is for tests that need to test interactivity (things like events, refs, etc.). The latter tests are only possible when `ReactDOM.render` has been called, which doesn't happen in the server render-only tests.

I tried to explain that in the comment above the function, but obviously it wasn't clear enough. Any suggestions on how to make it clearer?

spicyj commented on src/renderers/dom/shared/__tests__/ReactServerRendering-test.js:
```diff
@@ -23,20 +23,132 @@ var ReactTestUtils;
+  ReactReconcileTransaction = require('ReactReconcileTransaction');
+
+  ExecutionEnvironment = require('ExecutionEnvironment');
+  ExecutionEnvironment.canUseDOM = false;
```
This should be conditional on whether we are going to do client or server rendering, correct?

aickin replied:
You're right; this was an incantation I copied from the old code without thinking. I've changed it in a91c836 to change canUseDom to true before the call to `ReactDOM.render` in `renderToDom` (and change it back after). Seems to work right AFAICT.

spicyj commented on src/renderers/dom/shared/__tests__/ReactServerRendering-test.js:
```diff
@@ -549,4 +661,27 @@ describe('ReactDOMServer', () => {
   });
+
+  describe('basic rendering', function() {
+    itRenders('should render a blank div', render =>
```
nit: can we rename this function or the tests so that "it renders should render a blank div" is an actual sentence/phrase?

aickin replied:
No problem. Changed in a91c836.

spicyj commented on src/renderers/dom/shared/__tests__/ReactServerRendering-test.js:
```diff
@@ -549,4 +661,27 @@ describe('ReactDOMServer', () => {
+    );
+
+    itRenders('should render a self-closing tag', render =>
+      render(<br />).then(e => expect(e.tagName.toLowerCase()).toBe('br')));
```
(I think tagName is guaranteed to be uppercase so this could just be expect(e.tagName).toBe('BR')?)

aickin replied:
Good call. Modified all three tests in a91c836. Thanks!

aickin commented:
Thanks, @spicyj! I think I've addressed all the comments in your review in a91c836. I named the file `ReactDOMServerIntegration-test.js` because it's more of an integration-y test, but I'm not in love with the name at all; open to any suggestions.

I'm very interested in enabling async/await, but I'll put it in a separate PR so as to not slow this PR down.

Thank you for your time and thoughtful review!

spicyj commented:
Actually, for async/await – let's do the configuration in this file https://github.com/facebook/react/blob/3788654d57933461961a0296afd924493129e51d/scripts/jest/preprocessor.js#L29 and only do it if the file path matches `__tests__` so we don't accidentally introduce async/await into our actual bundle code. (Probably unlikely but I'd rather be safe.)

aickin commented:
> I think it would be nice to assert on the warning/error message text... Maybe it can just take expectedErrors as an array?

I was also thinking that's probably a good idea, or maybe assert on a substring/regex of the text. I'd be happy to do that a bit later.

aickin commented:
> Actually, for async/await – let's do the configuration in this file ...

Darn it; I almost had the async PR filed before you asked for this... ;)

Seriously though: good suggestion and thanks for it.

spicyj commented:
Sorry. :)
